components/brave_wallet/browser/zcash/zcash_create_orchard_to_orchard_transaction_task.cc

```diff
@@ -0,0 +1,149 @@
+  base::SequencedTaskRunner::GetCurrentDefault()->PostTask(
+      FROM_HERE,
+      base::BindOnce(&ZCashCreateOrchardToOrchardTransactionTask::WorkOnTask,
+                     weak_ptr_factory_.GetWeakPtr()));
+}
+
+void ZCashCreateOrchardToOrchardTransactionTask::WorkOnTask() {
+  if (error_) {
+    std::move(callback_).Run(base::unexpected(*error_));
+    zcash_wallet_service_->CreateTransactionTaskDone(this);
+    return;
+  }
+
+  if (!spendable_notes_) {
+    GetSpendableNotes();
+    return;
+  }
+
+  if (!transaction_) {
+    CreateTransaction();
+    return;
+  }
+
+  std::move(callback_).Run(base::ok(std::move(*transaction_)));
+  zcash_wallet_service_->CreateTransactionTaskDone(this);
+}
+
+void ZCashCreateOrchardToOrchardTransactionTask::GetSpendableNotes() {
+  context_.sync_state->AsyncCall(&OrchardSyncState::GetSpendableNotes)
+      .WithArgs(context_.account_id.Clone())
+      .Then(base::BindOnce(
+          &ZCashCreateOrchardToOrchardTransactionTask::OnGetSpendableNotes,
+          weak_ptr_factory_.GetWeakPtr()));
+}
+
+void ZCashCreateOrchardToOrchardTransactionTask::OnGetSpendableNotes(
+    base::expected<std::vector<OrchardNote>, OrchardStorage::Error> result) {
+  if (!result.has_value()) {
+    error_ = result.error().message;
+    ScheduleWorkOnTask();
+    return;
+  }
+
+  spendable_notes_ = result.value();
+  ScheduleWorkOnTask();
+}
+
+void ZCashCreateOrchardToOrchardTransactionTask::CreateTransaction() {
+  CHECK(spendable_notes_);
+  auto pick_result = PickZCashOrchardInputs(spendable_notes_.value(), amount_);
+  if (!pick_result) {
+    error_ = "Can't pick inputs";
+    ScheduleWorkOnTask();
+    return;
+  }
+
+  ZCashTransaction zcash_transaction;
+  for (const auto& note : pick_result.value().inputs) {
+    OrchardInput orchard_input;
+    orchard_input.note = note;
+    zcash_transaction.orchard_part().inputs.push_back(std::move(orchard_input));
+  }
+  zcash_transaction.set_fee(pick_result->fee);
+
+  // Create shielded change
+  if (pick_result->change != 0) {
+    OrchardOutput& orchard_output =
+        zcash_transaction.orchard_part().outputs.emplace_back();
+    orchard_output.value = pick_result->change;
+    auto change_addr =
+        zcash_wallet_service_->keyring_service_->GetOrchardRawBytes(
+            context_.account_id.Clone(),
+            mojom::ZCashKeyId::New(context_.account_id->account_index,
+                                   1 /* internal */, 0));
+    if (!change_addr) {
+      error_ = l10n_util::GetStringUTF8(IDS_WALLET_INTERNAL_ERROR);
+      ScheduleWorkOnTask();
+      return;
+    }
+    orchard_output.addr = *change_addr;
+  }
+
+  // Create shielded output
+  OrchardOutput& orchard_output =
+      zcash_transaction.orchard_part().outputs.emplace_back();
+  orchard_output.value = zcash_transaction.TotalInputsAmount() -
+                         zcash_transaction.fee() - pick_result->change;
+  orchard_output.addr = receiver_;
```

Comment:
Needs some explanation or guards so this does not overflow

Reply:
Done

Reply:
To avoid copy paste that should be based on base::CheckedNumeric

Reply:
I'm probably missing something, but how is this being done? Is it done via the typing system since it's declared as a const or is it meant to be handled by the caller in someway?